File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from collections import deque

N, M, V = map(int, sys.stdin.readline().split())

graph = [[False for _ in range(N + 1)] for _ in range(N + 1)]

for _ in range(M):
    v1, v2 = map(int, sys.stdin.readline().split())
    graph[v1][v2] = True
    graph[v2][v1] = True

visited = [False for _ in range(N + 1)]

need_visit = deque()
need_visit.append(V)

while need_visit:

    v = need_visit.popleft()
    if visited[v]:
        continue
    visited[v] = True
    print(v, end=' ')

    for i in range(N, 0, -1):
        if graph[v][i] and visited[i] == False:
            need_visit.appendleft(i)
print()


visited = [False for _ in range(N + 1)]

need_visit = deque()
need_visit.append(V)

while need_visit:

    v = need_visit.pop()
    if visited[v]:
        continue
    visited[v] = True
    print(v, end=' ')

    for i in range(1, N + 1):
        if graph[v][i] and visited[i] == False:
            need_visit.appendleft(i)
print()

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.cc
#include <iostream>
#include <queue>
#include <set>

using namespace std;

void dfs(const vector<set<int>> &graph, int N, int V){
    vector<int> need_visit;
    need_visit.push_back(V);

    vector<bool> visited(N + 1, false);

    while(!need_visit.empty()){

        int cur_node = need_visit.back();
        need_visit.pop_back();

        if (visited[cur_node])
            continue;

        cout << cur_node << " ";
        visited[cur_node] = true;

        vector<int> to_reverse;
        for(int next_node: graph[cur_node]){
            if(!visited[next_node])
                to_reverse.push_back(next_node);
        }

        while(!to_reverse.empty()){
            int reversed_node = to_reverse.back();
            to_reverse.pop_back();

            need_visit.push_back(reversed_node);
        }
    }
    cout << endl;
}

void bfs(const vector<set<int>> &graph, int N, int V){
    queue<int> need_visit;
    need_visit.push(V);

    vector<bool> visited(N + 1, false);

    while(!need_visit.empty()){
        int cur_node = need_visit.front();
        need_visit.pop();

        if(visited[cur_node])
            continue;

        cout << cur_node << " ";
        visited[cur_node] = true;

        for(int next_node: graph[cur_node]){
            if(!visited[next_node])
                need_visit.push(next_node);
        }
    }
}

int main(){
    int N, M, V;

    cin >> N >> M >> V;

    vector<set<int>> graph(N + 1);

    for(int i = 0; i < M; i++){
        int node1, node2;
        cin >> node1 >> node2;
        graph[node1].insert(node2);
        graph[node2].insert(node1);
    }

    dfs(graph, N, V);
    bfs(graph, N, V);
}